orchestrator/src/coordinator.rs

```diff
@@ -0,0 +1,622 @@
+                        if let Channel::Nightly = req.channel {
+                            args.push("-Z");
+                            args.push("asm-comments");
+                        }
+
+                        args.push("-C");
+                        match flavor {
+                            Att => args.push("llvm-args=-x86-asm-syntax=att"),
+                            Intel => args.push("llvm-args=-x86-asm-syntax=intel"),
+                        }
+                    }
+                    LlvmIr => args.extend(&["--", "--emit", "llvm-ir=compilation"]),
+                    Mir => args.extend(&["--", "--emit", "mir=compilation"]),
+                    Hir => args.extend(&["--", "-Zunpretty=hir", "-o", output_path]),
+                    Wasm => args.extend(&["-o", output_path]),
+                }
+                let mut envs = HashMap::new();
+                if req.backtrace {
+                    envs.insert("RUST_BACKTRACE".to_owned(), "1".to_owned());
+                }
+                batch.push(Request::ExecuteCommand(ExecuteCommandRequest {
+                    cmd: "cargo".to_owned(),
+                    args: args
+                        .into_iter()
+                        .map(|s| s.to_owned())
+                        .collect::<Vec<String>>(),
+                    envs,
+                    cwd: None,
+                }));
+                batch.push(Request::ReadFile(ReadFileRequest {
+                    path: output_path.to_owned(),
+                }));
+                Ok(Job { reqs: batch })
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum RequestKind {
+    Compile,
+}
+
+impl RequestKind {
+    fn kind(req: &HighLevelRequest) -> Self {
+        match req {
+            HighLevelRequest::Compile(_) => RequestKind::Compile,
+        }
+    }
+}
+
+impl From<(JobReport, RequestKind)> for HighLevelResponse {
+    fn from(value: (JobReport, RequestKind)) -> Self {
+        use crate::message::*;
+
+        let (JobReport { resps }, kind) = value;
+        let responses = resps;
+        match kind {
+            RequestKind::Compile => {
+                let mut success = false;
+                let mut code = String::new();
+                if let Some(ResponseResult(Ok(Response::ReadFile(ReadFileResponse(content))))) =
+                    responses.last()
+                {
+                    match from_utf8(content) {
+                        Ok(content) => {
+                            success = true;
+                            code = content.to_owned();
+                        }
+                        Err(_) => {
+                            success = false;
+                            code = "Error: Compilation output is not valid UTF-8 string".to_owned();
+                        }
+                    }
+                }
+                HighLevelResponse::Compile(CompileResponse {
+                    success,
```

Comment:
This is supposed to indicate if *running the process* was successful, not decoding UTF-8.